### PR TITLE
Add reading of specific images from multipage tiff

### PR DESCRIPTION
**Pull request by HarryDC**
The current implementation of `imreadmulti` reads all images from the multipage file, sometimes this is not the required functionality, for example due to memory contraints, or really just wanting to target a specific subrange of the multipage file. This PR allows any subsequence to be read . 

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by asenyaev** [comment]
jenkins cn please retry a build

**Comment by sturkmen72** [comment]
there is my [previous work](https://github.com/opencv/opencv/pull/17753) related

**Comment by asenyaev** [comment]
jenkins cn please retry a build

**Comment by asenyaev** [comment]
jenkins cn please retry a build

**Comment by HarryDC** [comment]
@sturkmen72 yes i saw that, i am trying a bit smaller approach that stays within the current interfaces.

**Comment by HarryDC** [comment]
@asenyaev it looks like the windows10 build is sufferring from some other issues 

```
[ RUN      ] videoio/videoio_bunny.frame_timestamp/11, where GetParam() = ("mp4", CV_MJPEG)
[     SKIP ] Backend CV_MJPEG can't open the video: c:/Slave/workspace/precommit/windows10\opencv_extra\testdata/highgui/video/big_buck_bunny.mp4
[       OK ] videoio/videoio_bunny.frame_timestamp/11 (0 ms)
[ RUN      ] videoio/videoio_bunny.frame_timestamp/12, where GetParam() = ("mpg", MSMF)
Cancelling nested steps due to timeout
Sending interrupt signal to process
script returned exit code -1
[Pipeline] echo
Aggregating results
[Pipeline] junit
Recording test results
No test report files were found. Configuration error?
[Pipeline] }
[Pipeline] // timeout
[Pipeline] }
[Pipeline] // stage
[Pipeline] }
[Pipeline] // dir
[Pipeline] }
[Pipeline] // withEnv
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
Setting status of c5834abaf578a1f82e0dd0fa5cde08e434797e4f to FAILURE with url https://build.opencv.org.cn/job/precommit/job/windows10/1450/ and message: 'Build finished. '
Using context: OpenCV CN Windows 10 x64
ERROR: No test report files were found. Configuration error?
Finished: FAILURE
```

**Comment by asenyaev** [comment]
@HarryDC yes, you are right, it stuck on the videoio testin stage. The CI server for **OpenCV CN Ubuntu 18.04 x86-64** build is down at this moment. I will restart builds for this PR as soon as it is online.

**Comment by asenyaev** [comment]
jenkins cn please retry a build

**Comment by asenyaev** [comment]
jenkins cn please retry a build

**Comment by asenyaev** [comment]
jenkins cn please retry a build

**Comment by asenyaev** [comment]
jenkins cn please retry a build

**Comment by asenyaev** [comment]
jenkins cn please retry a build

**Comment by HarryDC** [comment]
@asenyaev what are next steps ? Who could provide feedback for this ? 

Thanks

File: modules/imgcodecs/src/loadsave.cpp [hunk 1]
```diff
@@ -600,6 +600,116 @@ imreadmulti_(const String& filename, int flags, std::vector<Mat>& mats)
...
+            std::cerr << "imreadmulti_('" << filename << "'): can't read data: unknown exception" << std::endl << std::flush;
+        }
+        if (!success)
+            break;
+
+        // optionally rotate the data if EXIF' orientation flag says so
+        if ((flags & IMREAD_IGNORE_ORIENTATION) == 0 && flags != IMREAD_UNCHANGED)
+        {
+            ApplyExifOrientation(decoder->getExifTag(ORIENTATION), mat);
+        }
```
Comment (alalek) [inline modules/imgcodecs/src/loadsave.cpp:700]: Thank you for the contribution!

Main problem of this patch is large amount of copy-pasted code.
Lets try to reduce that.

1. Upgrade already existed implementation by adding new parameters
2. Add simple wrapper for old API.

```.diff
 static bool
-imreadmulti_(const String& filename, int flags, std::vector<Mat>& mats)
+imreadmulti_(const String& filename, int flags, std::vector<Mat>& mats, size_t start, size_t count)
{
    ... extra changes for new parameters ...
    ... support count=-1 as "nolimit" ...
}

+static
+bool imreadmulti_(const String& filename, int flags, std::vector<Mat>& mats)
+{
+    return imreadmulti_(filename, flags, mats, 0/*start*/, -1/*count*/);
+}
```

---

Prefer to keep changes minimal.

File: modules/imgcodecs/src/loadsave.cpp [hunk 2]
```diff
@@ -639,6 +749,90 @@ bool imreadmulti(const String& filename, std::vector<Mat>& mats, int flags)
...
+*
+*  This function merely calls the actual implementation above and returns itself.
+*
+* @param[in] filename File to load
+* @param[in] mats Reference to C++ vector<Mat> object to hold the images
+* @param[in] start Start index for the sequence to load
+* @param[in] count number of images to load
+* @param[in] flags Flags you wish to set.
+*
+*/
```
Comment (alalek) [inline modules/imgcodecs/src/loadsave.cpp:764]: BTW, Doxygen comments are not used in .cpp files. They can be safely omitted.

File: modules/imgcodecs/src/loadsave.cpp (hunk 1 above)
Comment (HarryDC) [reply modules/imgcodecs/src/loadsave.cpp:700]: I wouldn't usually use -1 converted into `size_t` if that is acceptable that of course works, i'll push something new

Comment (alalek) [reply modules/imgcodecs/src/loadsave.cpp:700]: Right, -1 is not a good value for `size_t`.

Consider using "int" type for scalars in OpenCV public API. `size_t` is not well supported by OpenCV bindings.

File: modules/imgcodecs/src/loadsave.cpp (hunk 2 above)
Comment (HarryDC) [reply modules/imgcodecs/src/loadsave.cpp]: Thanks for the feedback

File: modules/imgcodecs/src/loadsave.cpp (hunk 1 above)
Comment (HarryDC) [reply modules/imgcodecs/src/loadsave.cpp:602]: Changed all the params to int now ... included the other changes

File: modules/imgcodecs/src/loadsave.cpp [hunk 3]
```diff
@@ -495,44 +495,44 @@ imread_( const String& filename, int flags, Mat& mat )
...
 
     /// if no decoder was found, return nothing.
-    if (!decoder){
+    if (!decoder) {
         return 0;
     }
 
+    if (start < 0) {
+        return 0;
+    }
```
Comment (alalek) [inline modules/imgcodecs/src/loadsave.cpp:522]: Perhaps this should be an assertion check: `CV_CheckGE(start, 0, "");`

Also it make sense to move this validation check to the beginning of the function (before trying decoders).

File: modules/imgcodecs/src/loadsave.cpp [hunk 4]
```diff
@@ -636,9 +648,81 @@ bool imreadmulti(const String& filename, std::vector<Mat>& mats, int flags)
...
+    // read the header to make sure it succeeds
+    try
+    {
+        // read the header to make sure it succeeds
+        if (!decoder->readHeader())
+            return 0;
+    }
+    catch (const cv::Exception& e)
+    {
+        std::cerr << "imreadmulti_('" << filename << "'): can't read header: " << e.what() << std::endl << std::flush;
```
Comment (alalek) [inline modules/imgcodecs/src/loadsave.cpp:698]: > imreadmulti_

=> `imcount_`

**Comment by HarryDC** [comment]
jenkins cn please retry a build

**Review by alalek — APPROVED** [review]

**Pull request merged**